flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SameArgumentsInputTypeStrategy.java

```diff
@@ -23,37 +23,47 @@
 import org.apache.flink.table.types.DataType;
 import org.apache.flink.table.types.inference.ArgumentCount;
 import org.apache.flink.table.types.inference.CallContext;
-import org.apache.flink.table.types.inference.ConstantArgumentCount;
 import org.apache.flink.table.types.inference.InputTypeStrategy;
 import org.apache.flink.table.types.inference.Signature;
+import org.apache.flink.table.types.logical.LegacyTypeInformationType;
 import org.apache.flink.table.types.logical.LogicalType;
 import org.apache.flink.table.types.logical.utils.LogicalTypeGeneralization;
 import org.apache.flink.table.types.utils.TypeConversions;
 
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
 /**
- * {@link InputTypeStrategy} specific for {@link org.apache.flink.table.functions.BuiltInFunctionDefinitions#ARRAY}.
- *
- * <p>It expects at least one argument. All the arguments must have a common super type.
+ * An {@link InputTypeStrategy} that expects that all arguments have a common type.
  */
 @Internal
-public class ArrayInputTypeStrategy implements InputTypeStrategy {
+public class SameArgumentsInputTypeStrategy implements InputTypeStrategy {
+	private static final Signature.Argument COMMON_ARGUMENT = Signature.Argument.of("<COMMON>");
+	private final ArgumentCount argumentCount;
+
+	public SameArgumentsInputTypeStrategy(ArgumentCount argumentCount) {
+		this.argumentCount = argumentCount;
+	}
+
 	@Override
 	public ArgumentCount getArgumentCount() {
-		return ConstantArgumentCount.from(1);
+		return argumentCount;
 	}
 
 	@Override
 	public Optional<List<DataType>> inferInputTypes(
-		CallContext callContext,
-		boolean throwOnFailure) {
+			CallContext callContext,
+			boolean throwOnFailure) {
 		List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
-		if (argumentDataTypes.size() == 0) {
-			return Optional.empty();
+
+		if (callContext.getArgumentDataTypes()
+				.stream()
+				.anyMatch(dataType -> dataType.getLogicalType() instanceof LegacyTypeInformationType)) {
```

Comment:
convert to logical types at the beginning of this method to reuse them here and below